The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/10815984934/job/30006148427) [(plain)](https://github.com/rust-lang/rust/commit/ed8e7156e48e2438899bcabb17911fd36e2d066e/checks/30006148427/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---
#13 2.842 Building wheels for collected packages: reuse
#13 2.843   Building wheel for reuse (pyproject.toml): started
#13 3.090   Building wheel for reuse (pyproject.toml): finished with status 'done'
#13 3.091   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132715 sha256=dfa09868353292d98f811d3efdb0d54d07389e808efc71d68e3b93c514bf8bec
#13 3.091   Stored in directory: /tmp/pip-ephem-wheel-cache-6k7pm_09/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#13 3.094 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#13 3.490 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#13 3.491 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#13 4.027 Collecting virtualenv
#13 4.027 Collecting virtualenv
#13 4.077   Downloading virtualenv-20.26.4-py3-none-any.whl (6.0 MB)
#13 4.157      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 6.0/6.0 MB 76.8 MB/s eta 0:00:00
#13 4.221 Collecting filelock<4,>=3.12.2
#13 4.229   Downloading filelock-3.16.0-py3-none-any.whl (16 kB)
#13 4.250 Collecting distlib<1,>=0.3.7
#13 4.258   Downloading distlib-0.3.8-py2.py3-none-any.whl (468 kB)
#13 4.265      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 468.9/468.9 KB 91.9 MB/s eta 0:00:00
#13 4.300 Collecting platformdirs<5,>=3.9.1
#13 4.307   Downloading platformdirs-4.3.2-py3-none-any.whl (18 kB)
#13 4.388 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#13 4.577 Successfully installed distlib-0.3.8 filelock-3.16.0 platformdirs-4.3.2 virtualenv-20.26.4
#13 DONE 4.7s

#14 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#14 DONE 0.0s
---
DirectMap4k:      223168 kB
DirectMap2M:     8165376 kB
DirectMap1G:    10485760 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
    Finished `dev` profile [unoptimized] target(s) in 0.04s
##[endgroup]
downloading https://ci-artifacts.rust-lang.org/rustc-builds-alt/f7f8bdf2e03918cc3742db3bb37b7cfe9994dbec/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz
extracting /checkout/obj/build/cache/llvm-f7f8bdf2e03918cc3742db3bb37b7cfe9994dbec-true/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz to /checkout/obj/build/x86_64-unknown-linux-gnu/ci-llvm
---
fmt check
Diff in /checkout/src/bootstrap/src/core/build_steps/dist.rs:475:
             }
 
             let tool_builder_compiler = builder.compiler(builder.top_stage, builder.config.build);
-            maybe_install_wasm_component_ld(builder, tool_builder_compiler, compiler.host, &dst_dir);
-            maybe_install_llvm_bitcode_linker(builder, tool_builder_compiler, compiler.host, &dst_dir);
+            maybe_install_wasm_component_ld(
+                tool_builder_compiler,
+                compiler.host,
+                &dst_dir,
+            );
+            );
+            maybe_install_llvm_bitcode_linker(
+                builder,
+                tool_builder_compiler,
+                compiler.host,
+                &dst_dir,
+            );
             maybe_install_llvm_tools(builder, compiler.host, &dst_dir);
             // Man pages
             // Man pages
fmt error: Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/rustfmt/bin/rustfmt" "--config-path" "/checkout" "--edition" "2021" "--unstable-features" "--skip-children" "--check" "/checkout/compiler/rustc_trait_selection/src/error_reporting/infer/nice_region_error/different_lifetimes.rs" "/checkout/compiler/rustc_trait_selection/src/error_reporting/infer/nice_region_error/mod.rs" "/checkout/compiler/rustc_trait_selection/src/error_reporting/infer/sub_relations.rs" "/checkout/src/bootstrap/src/utils/job.rs" "/checkout/src/bootstrap/src/utils/cache.rs" "/checkout/src/bootstrap/src/utils/cc_detect.rs" "/checkout/src/bootstrap/src/utils/metrics.rs" "/checkout/src/bootstrap/src/utils/tarball.rs" "/checkout/src/bootstrap/src/utils/mod.rs" "/checkout/src/bootstrap/src/utils/shared_helpers/tests.rs" "/checkout/src/bootstrap/src/utils/render_tests.rs" "/checkout/src/bootstrap/src/utils/change_tracker/tests.rs" "/checkout/src/bootstrap/src/utils/change_tracker.rs" "/checkout/src/bootstrap/src/utils/shared_helpers.rs" "/checkout/src/bootstrap/src/utils/helpers.rs" "/checkout/compiler/rustc_trait_selection/src/error_reporting/traits/suggestions.rs" "/checkout/compiler/rustc_trait_selection/src/error_reporting/traits/on_unimplemented.rs" "/checkout/compiler/rustc_trait_selection/src/error_reporting/traits/overflow.rs" "/checkout/compiler/rustc_trait_selection/src/error_reporting/traits/mod.rs" "/checkout/compiler/rustc_trait_selection/src/error_reporting/traits/fulfillment_errors.rs" "/checkout/compiler/rustc_trait_selection/src/error_reporting/traits/ambiguity.rs" "/checkout/compiler/rustc_trait_selection/src/error_reporting/mod.rs" "/checkout/src/bootstrap/src/core/builder.rs" "/checkout/src/bootstrap/src/core/metadata.rs" "/checkout/src/bootstrap/src/core/sanity.rs" "/checkout/src/bootstrap/src/core/config/tests.rs" "/checkout/src/bootstrap/src/core/config/flags.rs" "/checkout/src/bootstrap/src/core/config/config.rs" "/checkout/src/bootstrap/src/core/config/mod.rs" "/checkout/src/bootstrap/src/core/download.rs" "/checkout/src/bootstrap/src/core/mod.rs" "/checkout/compiler/rustc_trait_selection/src/traits/wf.rs" "/checkout/compiler/rustc_trait_selection/src/traits/misc.rs" "/checkout/compiler/rustc_trait_selection/src/traits/fulfill.rs" "/checkout/compiler/rustc_trait_selection/src/traits/object_safety.rs" "/checkout/compiler/rustc_trait_selection/src/traits/normalize.rs" "/checkout/compiler/rustc_trait_selection/src/traits/project.rs" "/checkout/compiler/rustc_trait_selection/src/traits/structural_normalize.rs" "/checkout/compiler/rustc_trait_selection/src/traits/const_evaluatable.rs" "/checkout/src/bootstrap/src/core/build_steps/install.rs" "/checkout/src/bootstrap/src/core/build_steps/compile.rs" "/checkout/src/bootstrap/src/core/build_steps/suggest.rs" "/checkout/src/bootstrap/src/core/build_steps/clean.rs" "/checkout/src/bootstrap/src/core/build_steps/llvm.rs" "/checkout/src/bootstrap/src/core/build_steps/setup.rs" "/checkout/src/bootstrap/src/core/build_steps/clippy.rs" "/checkout/src/bootstrap/src/core/build_steps/setup/tests.rs" "/checkout/src/bootstrap/src/core/build_steps/vendor.rs" "/checkout/src/bootstrap/src/core/build_steps/dist.rs" "/checkout/src/bootstrap/src/core/build_steps/tool.rs" "/checkout/src/bootstrap/src/core/build_steps/format.rs" "/checkout/src/bootstrap/src/core/build_steps/toolstate.rs" "/checkout/src/bootstrap/src/core/build_steps/run.rs" "/checkout/src/bootstrap/src/core/build_steps/perf.rs" "/checkout/src/bootstrap/src/core/build_steps/synthetic_targets.rs" "/checkout/src/bootstrap/src/core/build_steps/mod.rs" "/checkout/src/bootstrap/src/core/build_steps/check.rs" "/checkout/src/bootstrap/src/core/build_steps/doc.rs" "/checkout/src/bootstrap/src/core/build_steps/test.rs" "/checkout/compiler/rustc_trait_selection/src/traits/query/method_autoderef.rs" "/checkout/compiler/rustc_trait_selection/src/traits/query/normalize.rs" "/checkout/compiler/rustc_trait_selection/src/traits/query/evaluate_obligation.rs" "/checkout/compiler/rustc_trait_selection/src/traits/query/mod.rs" "/checkout/src/bootstrap/src/utils/helpers/tests.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
  local time: Wed Sep 11 16:51:36 UTC 2024
  network time: Wed, 11 Sep 2024 16:51:36 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

--- a/src/bootstrap/src/core/build_steps/compile.rs
+++ b/src/bootstrap/src/core/build_steps/compile.rs
@@ -29,7 +29,7 @@ use crate::utils::helpers::{
     self, exe, get_clang_cl_resource_dir, get_closest_merge_base_commit, is_debug_info, is_dylib,
     symlink_dir, t, up_to_date,
 };
-use crate::{CLang, Compiler, DependencyType, GitRepo, Mode, LLVM_TOOLS};
+use crate::{CLang, Compiler, DependencyType, GitRepo, Mode};
 
 #[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
 pub struct Std {
@@ -1912,52 +1912,26 @@ impl Step for Assemble {
         // delegates to the `rust-lld` binary for linking and then runs
         // logic to create the final binary. This is used by the
         // `wasm32-wasip2` target of Rust.
-        if builder.tool_enabled("wasm-component-ld") {
-            let wasm_component_ld_exe =
-                builder.ensure(crate::core::build_steps::tool::WasmComponentLd {
-                    compiler: build_compiler,
-                    target: target_compiler.host,
-                });
-            builder.copy_link(
-                &wasm_component_ld_exe,
-                &libdir_bin.join(wasm_component_ld_exe.file_name().unwrap()),
-            );
-        }
+        dist::maybe_install_wasm_component_ld(
+            builder,
+            build_compiler,
+            target_compiler.host,
+            &libdir_bin,
+            false,
+        );
 
-        if builder.config.llvm_enabled(target_compiler.host) {
-            let llvm::LlvmResult { llvm_config, .. } =
-                builder.ensure(llvm::Llvm { target: target_compiler.host });
-            if !builder.config.dry_run() && builder.config.llvm_tools_enabled {
-                let llvm_bin_dir =
-                    command(llvm_config).arg("--bindir").run_capture_stdout(builder).stdout();
-                let llvm_bin_dir = Path::new(llvm_bin_dir.trim());
+        dist::maybe_install_llvm_tools(builder, target_compiler.host, &libdir_bin, false);
 
-                // Since we've already built the LLVM tools, install them to the sysroot.
-                // This is the equivalent of installing the `llvm-tools-preview` component via
-                // rustup, and lets developers use a locally built toolchain to
-                // build projects that expect llvm tools to be present in the sysroot
-                // (e.g. the `bootimage` crate).
-                for tool in LLVM_TOOLS {
-                    let tool_exe = exe(tool, target_compiler.host);
-                    let src_path = llvm_bin_dir.join(&tool_exe);
-                    // When using `download-ci-llvm`, some of the tools
-                    // may not exist, so skip trying to copy them.
-                    if src_path.exists() {
-                        builder.copy_link(&src_path, &libdir_bin.join(&tool_exe));
-                    }
-                }
-            }
-        }
+        let self_contained_bin_dir = libdir_bin.join("self-contained");
+        t!(fs::create_dir_all(&self_contained_bin_dir));
 
-        if builder.config.llvm_bitcode_linker_enabled {
-            let src_path = builder.ensure(crate::core::build_steps::tool::LlvmBitcodeLinker {
-                compiler: build_compiler,
-                target: target_compiler.host,
-                extra_features: vec![],
-            });
-            let tool_exe = exe("llvm-bitcode-linker", target_compiler.host);
-            builder.copy_link(&src_path, &libdir_bin.join(tool_exe));
-        }
+        dist::maybe_install_llvm_bitcode_linker(
+            builder,
+            build_compiler,
+            target_compiler.host,
+            &self_contained_bin_dir,
+            false,
+        );
 
         // Ensure that `libLLVM.so` ends up in the newly build compiler directory,
         // so that it can be found when the newly built `rustc` is run.

--- a/src/bootstrap/src/core/build_steps/dist.rs
+++ b/src/bootstrap/src/core/build_steps/dist.rs
@@ -473,11 +473,29 @@ impl Step for Rustc {
                     );
                 }
             }
-            if builder.tool_enabled("wasm-component-ld") {
-                let src_dir = builder.sysroot_libdir(compiler, host).parent().unwrap().join("bin");
-                let ld = exe("wasm-component-ld", compiler.host);
-                builder.copy_link(&src_dir.join(&ld), &dst_dir.join(&ld));
-            }
+
+            let builder_compiler =
+                builder.compiler_for(builder.top_stage, builder.config.build, compiler.host);
+
+            maybe_install_wasm_component_ld(
+                builder,
+                builder_compiler,
+                compiler.host,
+                &dst_dir,
+                true,
+            );
+
+            let self_contained_bin_dir = dst_dir.join("self-contained");
+            t!(fs::create_dir_all(&self_contained_bin_dir));
+            maybe_install_llvm_bitcode_linker(
+                builder,
+                builder_compiler,
+                compiler.host,
+                &self_contained_bin_dir,
+                true,
+            );
+
+            maybe_install_llvm_tools(builder, compiler.host, &dst_dir, true);
 
             // Man pages
             t!(fs::create_dir_all(image.join("share/man/man1")));
@@ -2092,6 +2110,85 @@ pub fn maybe_install_llvm_runtime(builder: &Builder<'_>, target: TargetSelection
     // statically.
     if builder.llvm_link_shared() {
         maybe_install_llvm(builder, target, &dst_libdir, false);
+    }
+}
+
+/// Maybe add LLVM tools to the rustc sysroot.
+pub fn maybe_install_llvm_tools(
+    builder: &Builder<'_>,
+    target: TargetSelection,
+    dst_dir: &Path,
+    dereference_symlinks: bool,
+) {
+    if builder.config.llvm_enabled(target) {
+        let llvm::LlvmResult { llvm_config, .. } = builder.ensure(llvm::Llvm { target });
+        if !builder.config.dry_run() && builder.config.llvm_tools_enabled {
+            let llvm_bin_dir =
+                command(llvm_config).arg("--bindir").run_capture_stdout(builder).stdout();
+            let llvm_bin_dir = Path::new(llvm_bin_dir.trim());
+
+            // Since we've already built the LLVM tools, install them to the sysroot.
+            // This is the equivalent of installing the `llvm-tools-preview` component via
+            // rustup, and lets developers use a locally built toolchain to
+            // build projects that expect llvm tools to be present in the sysroot
+            // (e.g. the `bootimage` crate).
+            for tool in LLVM_TOOLS {
+                let tool_exe = exe(tool, target);
+                let src_path = llvm_bin_dir.join(&tool_exe);
+                // When using `download-ci-llvm`, some of the tools
+                // may not exist, so skip trying to copy them.
+                if src_path.exists() {
+                    builder.copy_link_internal(
+                        &src_path,
+                        &dst_dir.join(&tool_exe),
+                        dereference_symlinks,
+                    );
+                }
+            }
+        }
+    }
+}
+
+/// Maybe add `llvm-bitcode-linker` to the rustc sysroot.
+pub fn maybe_install_llvm_bitcode_linker(
+    builder: &Builder<'_>,
+    builder_compiler: Compiler,
+    target: TargetSelection,
+    dst_dir: &Path,
+    dereference_symlinks: bool,
+) {
+    if builder.config.llvm_bitcode_linker_enabled {
+        let llvm_bitcode_linker_exe = builder.ensure(tool::LlvmBitcodeLinker {
+            compiler: builder_compiler,
+            target,
+            extra_features: vec![],
+        });
+
+        builder.copy_link_internal(
+            &llvm_bitcode_linker_exe,
+            &dst_dir.join(llvm_bitcode_linker_exe.file_name().unwrap()),
+            dereference_symlinks,
+        );
+    }
+}
+
+/// Maybe add `wasm-component-ld` to the rustc sysroot.
+pub fn maybe_install_wasm_component_ld(
+    builder: &Builder<'_>,
+    builder_compiler: Compiler,
+    target: TargetSelection,
+    dst_dir: &Path,
+    dereference_symlinks: bool,
+) {
+    if builder.tool_enabled("wasm-component-ld") {
+        let wasm_component_ld_exe =
+            builder.ensure(tool::WasmComponentLd { compiler: builder_compiler, target });
+
+        builder.copy_link_internal(
+            &wasm_component_ld_exe,
+            &dst_dir.join(wasm_component_ld_exe.file_name().unwrap()),
+            dereference_symlinks,
+        );
     }
 }
 

